File: android/app/src/main/java/com/propertyfinder1/util/IntervalTree.java
package com.propertyfinder1.util;

import android.util.Log;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Map;
import java.util.List;

public class IntervalTree {
    private IntervalNode root;
    private Map<Interval, IntervalNode> currentIntervals;

    public IntervalTree() {
        root = null;
    }

    public void add(Interval interval) {
        root = insert(root, interval);
    }

    public void remove(Interval interval) {
        root = delete(root, interval);
    }

    public ArrayList<Interval> inorder() {
        ArrayList<Interval> intervalList = new ArrayList<>();
        inorder(root, intervalList);
        return intervalList;
    }

    public ArrayList<Interval> searchAll(int start, int end) {
        Interval searchInterval = new Interval(start, end);
        ArrayList<Interval> overlappingIntervals = new ArrayList<>();
        search(root, searchInterval, overlappingIntervals);
        return overlappingIntervals;
    }

    private IntervalNode newNode(Interval interval) {
        return new IntervalNode(interval);
    }

    private void inorder(IntervalNode node, ArrayList<Interval> intervalList) {
        if (node == null) {
            return;
        }
        inorder(node.getLeft(), intervalList);
        intervalList.add(node.getInterval());
        inorder(node.getRight(), intervalList);
    }

    private void search(IntervalNode node, Interval searchInterval, ArrayList<Interval> overlappingIntervals) {
        if (node == null) {
            return;
        }
        Interval currentInterval = node.getInterval();
        int cmp = searchInterval.compareTo(currentInterval);
        if (searchInterval.intersects(currentInterval)) {
            if (currentInterval.contains(searchInterval)) {
                overlappingIntervals.add(currentInterval);
            } else if (searchInterval.contains(currentInterval)) {
                search(node.getLeft(), searchInterval, overlappingIntervals);
                overlappingIntervals.add(currentInterval);
                search(node.getRight(), searchInterval, overlappingIntervals);
            } else if (cmp < 0) {
                search(node.getLeft(), searchInterval, overlappingIntervals);
                overlappingIntervals.add(currentInterval);
            } else {
                overlappingIntervals.add(currentInterval);
                search(node.getRight(), searchInterval, overlappingIntervals);
            }
        } else {
            if (cmp < 0) {
                search(node.getLeft(), searchInterval, overlappingIntervals);
            } else {
                search(node.getRight(), searchInterval, overlappingIntervals);
            }
        }
    }

    private IntervalNode insert(IntervalNode node, Interval interval) {
        if (node == null) {
            return newNode(interval);
        } else if (node.getInterval().equals(interval)) {
            return node;
        }
        int cmp = interval.compareTo(node.getInterval());
        if (cmp < 0) {
            node.setLeft(insert(node.getLeft(), interval));
        } else {
            node.setRight(insert(node.getRight(), interval));
        }
        return node;
    }

    private IntervalNode inorderSuccesor(IntervalNode root) {
        IntervalNode current = root.getRight();
        while (current !=  null && current.getLeft() != null) {
            current = current.getLeft();
        }
        return current;
    }

    private IntervalNode delete(IntervalNode node, Interval interval) {
        if(node == null) {
            return null;
        }
        int cmp = interval.compareTo(node.getInterval());
        if (cmp < 0) {
            node.setLeft(delete(node.getLeft(), interval));
        } else if (cmp > 0) {
            node.setRight(delete(node.getRight(), interval));
        } else {
            if (node.getLeft() == null) {
                return node.getRight();
            } else if(node.getRight() == null) {
                return node.getLeft();
            }
            IntervalNode successor = inorderSuccesor(node);
            node.setInterval(successor.getInterval());
            node.setRight(delete(node.getRight(), successor.getInterval()));
        }
        return node;
    }

}

File: ios/IntervalTree.swift
//
//  IntervalTree.swift
//  PropertyFinder1
//
//  Created by K Arun Kumar Reddy on 19/12/19.
//  Copyright © 2019 Facebook. All rights reserved.
//

import Foundation

class IntervalTree {
    private var root: IntervalNode?;
    
    init() {
        self.root = nil;
    }
    
    func add(interval: Interval) {
        root = insert(root: root, interval: interval);
    }
    
    func remove(interval: Interval) {
        root = delete(root: root, interval: interval);
    }
    
    func searchAll(start: Int, end: Int) -> Array<Interval> {
        let searchInterval: Interval = Interval(start: start, end: end);
        var overlappingIntervals: Array<Interval> = Array();
        search(root: root, searchInterval: searchInterval, overlappingIntervals: &overlappingIntervals);
        return overlappingIntervals;
    }
    
    func inorder() -> Array<Interval> {
        var intervalList: Array<Interval> = Array();
        inorder(root: root, intervalList: &intervalList);
        return intervalList;
    }
    
    private func newNode(interval: Interval) -> IntervalNode {
        return IntervalNode(interval: interval);
    }
    
    private func search(root: IntervalNode?, searchInterval: Interval, overlappingIntervals: inout Array<Interval>) {
        if let node = root {
            let currentInterval: Interval = node.getInterval();
            let cmp: Int = searchInterval.compareTo(interval: currentInterval);
            if (searchInterval.intersects(interval: currentInterval)) {
                if (currentInterval.contains(interval: searchInterval)) {
                    overlappingIntervals.append(currentInterval);
                } else if (searchInterval.contains(interval: currentInterval)) {
                    search(root: node.getLeft(), searchInterval: searchInterval, overlappingIntervals: &overlappingIntervals);
                    overlappingIntervals.append(currentInterval);
                    search(root: node.getRight(), searchInterval: searchInterval, overlappingIntervals: &overlappingIntervals);
                } else if (cmp < 0) {
                    search(root: node.getLeft(), searchInterval: searchInterval, overlappingIntervals: &overlappingIntervals);
                    overlappingIntervals.append(currentInterval);
                } else {
                    overlappingIntervals.append(currentInterval);
                    search(root: node.getRight(), searchInterval: searchInterval, overlappingIntervals: &overlappingIntervals);
                }
            } else {
                if (cmp < 0) {
                    search(root: node.getLeft(), searchInterval: searchInterval, overlappingIntervals: &overlappingIntervals);
                } else {
                    search(root: node.getRight(), searchInterval: searchInterval, overlappingIntervals: &overlappingIntervals);
                }
            }
        }
    }
    
    private func inorder(root: IntervalNode?, intervalList: inout Array<Interval>) {
        if let node = root {
            inorder(root: node.getLeft(), intervalList: &intervalList);
            intervalList.append(node.getInterval());
            inorder(root: node.getRight(), intervalList: &intervalList);
        }
    }
    
    private func insert(root: IntervalNode?, interval: Interval) -> IntervalNode? {
        if let node = root {
            if (node.getInterval() == interval) {
                return node;
            }
            let cmp = interval.compareTo(interval: node.getInterval());
            if (cmp < 0) {
                node.setLeft(node: insert(root: node.getLeft(), interval: interval));
            } else {
                node.setRight(node: insert(root: node.getRight(), interval: interval));
            }
            return root;
        } else {
            return newNode(interval: interval);
        }
    }
    
    private func delete(root: IntervalNode?, interval: Interval) -> IntervalNode? {
        if let node = root {
            let cmp = interval.compareTo(interval: node.getInterval());
            if (cmp < 0) {
                node.setLeft(node: delete(root: node.getLeft(), interval: interval));
            } else if (cmp > 0) {
                node.setRight(node: delete(root: node.getRight(), interval: interval));
            } else {
                if (node.getLeft() == nil) {
                    return node.getRight();
                } else if (node.getRight() == nil) {
                    return node.getLeft();
                }
                let successor = inorderSuccssor(root: node);
                node.setInterval(interval: successor.getInterval());
                node.setRight(node: delete(root: node.getRight(), interval: successor.getInterval()));
            }
            return node;
        } else {
            return nil;
        }
    }
    
    private func inorderSuccssor(root: IntervalNode) -> IntervalNode {
        if var current = root.getRight() {
            while (true) {
                if let left = current.getLeft() {
                    current = left;
                } else {
                    break;
                }
            }
            return current;
        }
        return root;
    }
}
